proposal/0000-record-dot-syntax.md

```diff
@@ -12,77 +12,276 @@ the link, and delete this bold sentence.**
 
 ## Proposed Change Specification
 
-Specify the change in precise, comprehensive yet concise language. Avoid words
-like "should" or "could". Strive for a complete definition. Your specification
-may include,
+For this extension we focus on the changes to the parsing rules, and the desugaring, with the expectation the type checking and renamer follow directly from that. To confirm these changes integrate as expected we have written [a prototype implementation](https://gitlab.haskell.org/shayne-fletcher-da/ghc/commits/record-dot-syntax) that parses and desugars the forms directly in the parser.
+
+### `RecordDotSyntax` language extension
+
+This change adds a new language extension (enabled at source via `{-# LANGUAGE RecordDotSyntax #-}` or on the command line via the flag `-XRecordDotSyntax`).
+
+When `RecordDotSyntax` is in effect, the use of '.' to denote function composition is disambiguated from the use of '.' to denote record field access by the presence (, respectively absence,) of whitespace trailing the '.'.
+
+Suppose the following datatype declarations.
+
+```haskell
+data Foo = Foo {foo :: Bar}
+data Bar = Bar {bar :: Baz}
+data Baz = Baz {baz :: Quux}
+data Quux = Quux {quux :: Int}
+```
+
+The existence of the builtin `HasField` typeclass means that it is possible to write code for getting and setting record fields like this:
+
+```haskell
+getQuux :: Foo -> Int
+getQuux a = getField @"quux" (getField @"baz" (getField @"bar" (getField @"foo" a)))
+
+setQuux :: Foo -> Int -> Foo
+setQuux a i = setField@"foo" a (setField@"bar" (getField @"foo" a) (setField@"baz" (getField @"bar" (getField @"foo" a)) (setField@"quux" (getField @"baz" (getField @"bar" (getField @"foo" a))) i)))
+```
+
+`RecordDotSyntax` enables new concrete syntax so that the following program is equivalent.
+
+```haskell
+getQuux a = a.foo.bar.baz.quux
+setQuux a i = a{foo.bar.baz.quux = i}
+```
+
+In the event the language extension is enabled:
+
+| Expression | Equivalent |
+| -- | -- |
+| `e.lbl` | `getField @"lbl" e` the `.` cannot have whitespace on either side |
+| `e{lbl = val}` | `setField @"lbl" e val` |
+| `(.lbl)` | `(\x -> x.lbl)` the `.` cannot have whitespace after |
+| `e{lbl1.lbl2 = val}` | `e{lbl1 = (e.lbl1){lbl2 = val}}` performing a nested update |
+| `e{lbl * val}` | `e{lbl = e.lbl * val}` where `*` can be any operator |
+| `e{lbl1.lbl2}` | `e{lbl1.lbl2 = lbl2}` when punning is enabled|
```

Comment:
How does this play with record wildcards? Presumably `e{lbl1.lbl2...}` doesn't work. I could imagine that `e.lbl1.lbl2{..}` might work, in which case `e{lbl1.lbl2 = e.lbl1.lbl2{..}}` would also work.

If you supported nested updates like `e{lbl1.lbl2{lbl3 = foo}}`, maybe `e{lbl1.lbl2{..}}` would allow updating all fields in `e.lbl1.lbl2` for which I have something in scope. Are such nested updates included in the proposal?

Reply:
You can't use `{..}` in a record update only record construction, so you'd have to do `C{lbl1.lbl2,..}`, which means `C{lbl2=lbl1.lbl2,..}` with the standard meaning. Putting `{..}` at the end of a selector isn't a valid place for a pun today, so remains invalid.